index.js

```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
+                }
+
+                if (!upcomingInvoice.error && upcomingInvoice.created) {
+                    posthog.capture('Upcoming Invoice', {
+                        amount: invoiceData.amount_due,
+                        invoice_date: new Date(upcomingInvoiceDate).toLocaleDateString('en-GB'),
+                        product: invoiceData.product,
+                        quantity: invoiceData.quantity,
+                        stripe_customer_id: customer.id,
+                        distinct_id: customer.email || customer.id,
+                        $set: customer.email ? { email: customer.email } : undefined
+                    })
                 }
             }
         }
 
-        const customerRecordExists = await storage.get(customer.id)
+        // Note: I think if we want to automatically update users we should just update properties instead of firing events
+        // I also think we should perhaps restrict the properties we send so it doesn't feel overwhelming to
+        // see a lot of properties you may not understand.
+
+        // const customerRecordExists = await storage.get(customer.id)
+        //
+        // if (!customerRecordExists) {
+        //     await storage.set(customer.id, true)
+        // }
+        //
+        // if (global.onlyRegisterNewCustomers && customerRecordExists) {
+        //     break
+        // }
+        //
+        // posthog.capture(customerRecordExists ? 'update_stripe_customer' : 'new_stripe_customer', {
+        //     ...properties,
+        //     $set: basicProperties
+        // })
+    }
 
-        if (!customerRecordExists) {
-            await storage.set(customer.id, true)
-        }
+    logAggregatedInvoices(invoicesByProduct)
 
-        if (global.onlyRegisterNewCustomers && customerRecordExists) {
-            break
-        }
+    await cache.set('_lastRun', new Date().getTime())
+}
 
-        posthog.capture(customerRecordExists ? 'update_stripe_customer' : 'new_stripe_customer', {
-            ...properties,
-            $set: basicProperties
+function logAggregatedInvoices(invoicesByProduct) {
+    const totalsByProduct = {}
+    // First we need to sum all the invoices for each product.
+    for (const [product, invoices] of Object.entries(invoicesByProduct)) {
+        invoices.forEach((invoice) => {
+            if (product in totalsByProduct) {
+                totalsByProduct[product] += invoice.amount_due
+            } else {
+                totalsByProduct[product] = invoice.amount_due
+            }
         })
     }
-
-    await cache.set('_lastRun', new Date().getTime())
+    // Now for each product type, we send an event to PostHog with the sum total for that product.
+    for (const [product, billingSum] of Object.entries(totalsByProduct)) {
+        const props = {
+            amount: parseFloat(billingSum.toFixed(2)),
+            product: product,
+            distinct_id: 'posthog-stripe-plugin'
```

Comment:
This isn't necessary - events with no `distinct_id` set will automatically be associated with the "plugin user"